core/src/main/java/org/apache/calcite/sql/parser/SqlParserUtil.java

```diff
@@ -738,17 +741,7 @@ public static ParsedCollation parseCollation(String in) {
     }
 
     Charset charset = SqlUtil.getCharset(charsetStr);
-    String[] localeParts = localeStr.split("_");
-    Locale locale;
-    if (1 == localeParts.length) {
-      locale = new Locale(localeParts[0]);
-    } else if (2 == localeParts.length) {
-      locale = new Locale(localeParts[0], localeParts[1]);
-    } else if (3 == localeParts.length) {
-      locale = new Locale(localeParts[0], localeParts[1], localeParts[2]);
-    } else {
-      throw RESOURCE.illegalLocaleFormat(localeStr).ex();
```

Comment:
It seems that after these changes we are never going to throw the `illegalLocaleFormat` exception. Are we going to throw another exception? Are we going to fallback silently to something else? 

If we cannot keep the old behavior then we should document the breaking change in `history.md` file. Also if we never throw `RESOURCE.illegalLocaleFormat` then we maybe we should also drop the respective (dead) code.

Reply:
Since this change makes use of IETF BCP 47 language tag string (used by Locale#forLanguageTag) there could be more parts. Fallback is jdk's fallback

Thanks about mentioning `RESOURCE.illegalLocaleFormat`

Reply:
First it says (in jdk19 javadoc) to look for `Obtaining Locales` section at https://github.com/openjdk/jdk/commit/526e73498eef5c7608845501ab4ebef0997a5c0d#diff-bae63085ee0310e91011b048b2b59d0986eedf702dee10d7b0ab75c8e20a61eaR728
and then in that section there are 2 ways as mentioned at https://github.com/openjdk/jdk/commit/526e73498eef5c7608845501ab4ebef0997a5c0d#diff-bae63085ee0310e91011b048b2b59d0986eedf702dee10d7b0ab75c8e20a61eaR245
1. `Locale#forLanguageTag` which is used in this PR
2. `Locale#of` which is available only since jdk19 (https://github.com/openjdk/jdk/pull/7947).

>Do all Calcite unit tests run if we use setLanguageTag?

I replaced all occurrences of `Locale` constructor (including tests) then yes...
Or do you mean something else?

 >Are we sure that we do not break some SQL standard specification around COLLATE clause (etc.) by doing this change?
 
 That thing unfortunately I don't know how to check...

Reply:
Before the changes in this PR there was somewhat strict control on what formats are allowed/supported when working with locales and an exception was thrown if the requested locale is now invalid.

Now with the use of `Locale#forLanguageTag` we will never get an error since it appears that we will always fallback to some defaults. 

Instead of using the `Locale#forLanguageTag` we could use the `Locale.Builder.setLanguageTag`, which as far as I understand is not as permissive,  and will thrown an error when the locale is not well formed: https://download.java.net/java/early_access/panama/docs/api/java.base/java/util/Locale.Builder.html#setLanguageTag(java.lang.String)

In the context of this PR I was wondering if the `Locale.Builder.setLanguageTag` is a better alternative to `Locale#forLanguageTag`.

The latest standard can be found here: https://www.iso.org/standard/76584.html 
It is not free but you may find older versions available online (https://modern-sql.com/standard). The Part 2: Foundations covers the `COLLATE` clause.

Reply:
I'm ok with `Locale.Builder.setLanguageTag`
the only thing we should probably handle is that current behavior: in case of empty `localeString` it returns `Locale.ROOT` while `Locale.Builder.setLanguageTag` fails

Also added actual `localeStr` in exception message since `IllformedLocaleException` throwing by `setLamguageTag` does not contain it